wasm/names_test.go

```diff
@@ -0,0 +1,272 @@
+package wasm
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestCustomNameSection_EncodeData(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    *CustomNameSection
+		expected []byte
+	}{
+		{
+			name:  "empty",
+			input: &CustomNameSection{FunctionNames: map[uint32]string{}, LocalNames: map[uint32]map[uint32]string{}},
+		},
+		{
+			name: "only module",
+			// Ex. (module $simple )
+			input: &CustomNameSection{ModuleName: "simple"},
+			expected: []byte{
+				0x00, // Module subsection ID zero
+				0x07, // 7 bytes to follow
+				0x06, // the Module name simple is 6 bytes long
+				's', 'i', 'm', 'p', 'l', 'e',
+			},
+		},
+		{
+			name: "module and function name",
+			//	(module $simple
+			//		(import "" "Hello" (func $hello))
+			//		(start $hello)
+			//	)
+			input: &CustomNameSection{
+				ModuleName:    "simple",
+				FunctionNames: map[uint32]string{0x00: "hello"},
+			},
+			expected: []byte{
+				0x00, // Module subsection ID zero
+				0x07, // 7 bytes to follow
+				0x06, // the Module name simple is 6 bytes long
+				's', 'i', 'm', 'p', 'l', 'e',
+				0x01, // function subsection ID one
+				0x08, // 8 bytes to follow
+				0x01, // one function name
+				0x00, // the function index is zero
+				0x05, // the function name hello is 5 bytes long
+				'h', 'e', 'l', 'l', 'o',
+			},
+		},
+		{
+			name: "two function names", // Ex. TinyGo which at one point didn't set a module name
+			//	(module
+			//		(import "wasi_snapshot_preview1" "args_sizes_get" (func $runtime.args_sizes_get (param i32, i32) (result i32)))
+			//		(import "wasi_snapshot_preview1" "fd_write" (func $runtime.fd_write (param i32, i32, i32, i32) (result i32)))
+			//	)
+			input: &CustomNameSection{
+				FunctionNames: map[uint32]string{
+					0x00: "runtime.args_sizes_get",
+					0x01: "runtime.fd_write",
+				},
+			},
+			expected: []byte{
+				0x01, // function subsection ID one
+				0x2b, // 43 bytes to follow
+				0x02, // two function names
+				0x00, // the function index is zero
+				0x16, // the function name runtime.args_sizes_get is 22 bytes long
+				'r', 'u', 'n', 't', 'i', 'm', 'e', '.', 'a', 'r', 'g', 's', '_', 's', 'i', 'z', 'e', 's', '_', 'g', 'e', 't',
+				0x01, // the function index is one
+				0x10, // the function name runtime.fd_write is 16 bytes long
+				'r', 'u', 'n', 't', 'i', 'm', 'e', '.', 'f', 'd', '_', 'w', 'r', 'i', 't', 'e',
+			},
+		},
+		{
+			name: "function with local names",
+			//	(module
+			//		(import "Math" "Mul" (func $mul (param $x f32) (param $y f32) (result f32)))
+			//		(import "Math" "Add" (func $add (param $l f32) (param $r f32) (result f32)))
+			//	)
+			input: &CustomNameSection{
+				FunctionNames: map[uint32]string{
+					0x00: "mul",
+					0x01: "add",
+				},
+				LocalNames: map[uint32]map[uint32]string{
+					0x00: {0x00: "x", 0x01: "y"},
+					0x01: {0x00: "l", 0x01: "r"},
+				},
+			},
+			expected: []byte{
```

Comment:
this is hard to compare against something else because wasm-tools doesn't encode locals properly and wabt doesn't display nice. However, if you take a best-of you can verify this manually against something besides the spec. Encode the example with `wat2wasm --debug-names foo.wat`, then print it with `wasm-tools dump foo.wasm`

```bash
$ wasm-tools dump foo.wasm
--snip--
0x0031 | 01 0b       | function names
0x0033 | 02          | 2 count
0x0034 | 00 03 6d 75 | Naming { index: 0, name: "mul" }
       | 6c         
0x0039 | 01 03 61 64 | Naming { index: 1, name: "add" }
       | 64         
0x003e | 02 11       | local names
0x0040 | 02          | 2 count
0x0041 | 00          | function 0 locals
0x0042 | 02          | 2 count
0x0043 | 00 01 78    | Naming { index: 0, name: "x" }
0x0046 | 01 01 79    | Naming { index: 1, name: "y" }
0x0049 | 01          | function 1 locals
0x004a | 02          | 2 count
0x004b | 00 01 6c    | Naming { index: 0, name: "l" }
0x004e | 01 01 72    | Naming { index: 1, name: "r" }